pkg/commands/bucket_validation.go

```diff
@@ -0,0 +1,302 @@
+	}
+
+	b.bucketClient = &retryingBucketClient{
+		Bucket:  bucketClient,
+		retries: b.retriesOnError,
+	}
+
+	for testRun := 0; testRun < b.testRuns; testRun++ {
+		err = b.createTestObjects(ctx)
+		if err != nil {
+			return errors.Wrap(err, "error when uploading test data")
+		}
+
+		err = b.validateTestObjects(ctx)
+		if err != nil {
+			return errors.Wrap(err, "error when validating test data")
+		}
+
+		err = b.deleteTestObjects(ctx)
+		if err != nil {
+			return errors.Wrap(err, "error when deleting test data")
+		}
+
+		level.Info(b.logger).Log("testrun_successful", testRun+1)
+	}
+
+	return nil
+}
+
+func (b *BucketValidationCommand) printBucketConfigHelp() {
+	fs := flag.NewFlagSet("bucket-config", flag.ContinueOnError)
+	b.cfg.RegisterFlags(fs)
+
+	fmt.Fprintf(fs.Output(), `
+The following help text describes the arguments
+which may be specified in the string that gets
+passed to "-bucket-config".
+
+Example:
+cortextool bucket-validation --bucket-config='-backend=s3 -s3.endpoint=localhost:9000 -s3.bucket-name=example-bucket'
+
+`)
+	fs.Usage()
+}
+
+func (b *BucketValidationCommand) parseBucketConfig() error {
+	fs := flag.NewFlagSet("bucket-config", flag.ContinueOnError)
+	b.cfg.RegisterFlags(fs)
+	err := fs.Parse(strings.Split(b.bucketConfig, " "))
+	if err != nil {
+		return err
+	}
+
+	return b.cfg.Validate()
+}
+
+func (b *BucketValidationCommand) report(phase string, completed int) {
+	if completed == 0 || completed%b.reportEvery == 0 {
+		level.Info(b.logger).Log("phase", phase, "completed", completed, "total", b.objectCount)
+	}
+}
+
+func (b *BucketValidationCommand) setObjectNames() {
+	b.objectNames = make(map[string]string, b.objectCount)
+	for objectIdx := 0; objectIdx < b.objectCount; objectIdx++ {
+		b.objectNames[fmt.Sprintf("%s/%05X/", b.prefix, objectIdx)] = "testfile"
+	}
+}
+
+func (b *BucketValidationCommand) createTestObjects(ctx context.Context) error {
+	iteration := 0
+	for dirName, objectName := range b.objectNames {
+		b.report("creating test objects", iteration)
+		iteration++
+
+		objectPath := dirName + objectName
+		err := b.bucketClient.Upload(ctx, objectPath, strings.NewReader(b.objectContent))
+		if err != nil {
+			return errors.Wrapf(err, "failed to upload object (%s)", objectPath)
+		}
+
+		exists, err := b.bucketClient.Exists(ctx, objectPath)
+		if err != nil {
+			return errors.Wrapf(err, "failed to check if obj exists (%s)", objectPath)
+		}
+		if !exists {
+			return errors.Errorf("Expected obj %s to exist, but it did not", objectPath)
+		}
+	}
+	b.report("creating test objects", iteration)
+
+	return nil
+}
+
+func (b *BucketValidationCommand) validateTestObjects(ctx context.Context) error {
+	foundDirs := make(map[string]struct{}, b.objectCount)
+
+	level.Info(b.logger).Log("phase", "listing test objects")
+
+	err := b.bucketClient.Iter(ctx, b.prefix, func(dirName string) error {
+		foundDirs[dirName] = struct{}{}
+		return nil
+	})
+	if err != nil {
```

Comment:
Do you think it makes sense to compare the count of found test objects here?

Or at least add a output how many objects have been found. I just wondered what the log message is trying to tell only looking at the output.

Reply:
ic, that's a good idea, I'll add that. thx

Reply:
I added the object count validation in two places (42bcb98882f873a80dec6c94fd930fecdc2b220b):
1) when we do the initial list in `validateTestObjects()`
2) right after each delete in `deleteTestObjects()`